File: client/src/api/compliance.ts
// Libro de compras + resumen IVA (F29) — wrappers sobre los comandos Tauri
// (client/src-tauri/src/commands/compliance.rs). El libro se arma con las OC
// recepcionadas del período; mientras el operador no capture la factura real del
// proveedor, neto/IVA se derivan del total (19%) y la fila queda `declared:false`.
import { invoke } from "@tauri-apps/api/core";

/** Una fila del libro: un documento de proveedor. Money STRING. */
export interface PurchaseBookRow {
  purchase_order: string;
  /** Tipo DTE: 33 factura afecta, 34 exenta, 61 NC, 56 ND. */
  tipo: number;
  folio?: string | null;
  supplier_name: string;
  supplier_rut?: string | null;
  date: string;
  neto: string;
  iva: string;
  total: string;
  /** `false` = derivado del total; falta capturar la factura real. */
  declared: boolean;
}

export interface PurchaseBook {
  period: string;
  rows: PurchaseBookRow[];
  total_neto: string;
  total_iva: string;
  total: string;
  pending_declaration: number;
}

/** Débito (ventas) − crédito (compras) = IVA a pagar del período. */
export interface IvaSummary {
  period: string;
  iva_debito: string;
  iva_credito: string;
  /** Positivo = a pagar; negativo = remanente a favor. */
  iva_a_pagar: string;
  ventas_neto: string;
  compras_neto: string;
}

/** GET /api/v1/reports/libro-compras (admin+). Sin `period` → mes en curso. */
export function libroCompras(serverUrl: string, period?: string): Promise<PurchaseBook> {
  return invoke<PurchaseBook>("libro_compras", { serverUrl, period });
}

/** GET /api/v1/reports/iva (admin+). Sin `period` → mes en curso. */
export function ivaSummary(serverUrl: string, period?: string): Promise<IvaSummary> {
  return invoke<IvaSummary>("iva_summary", { serverUrl, period });
}

/** PATCH /api/v1/purchase-orders/{id}/factura (admin+) — capturar el documento
 *  del proveedor. Devuelve el libro del período ya actualizado. */
export function setPoInvoice(
  serverUrl: string,
  id: string,
  invoice: {
    folio?: string;
    date?: string;
    tipo?: number;
    neto?: string;
    iva?: string;
    total?: string;
  },
): Promise<PurchaseBook> {
  return invoke<PurchaseBook>("set_po_invoice", { serverUrl, id, ...invoice });
}

File: client/src-tauri/src/commands/compliance.rs
//! Libro de compras + resumen IVA (F29) y captura de la factura del proveedor.
//! Todo admin+ server-side (son cifras tributarias del negocio).

use secrecy::ExposeSecret;
use tauri::State;

use crate::http::{base, client, conn_error, error_message};
use crate::state::{token_of, SessionState};
use crate::types::{IvaSummary, PurchaseBook};

/// GET `/api/v1/reports/libro-compras?period=YYYY-MM` (Bearer, admin+).
/// Sin `period` el server usa el mes en curso.
#[tauri::command]
pub async fn libro_compras(
    state: State<'_, SessionState>,
    server_url: String,
    period: Option<String>,
) -> Result<PurchaseBook, String> {
    let token = token_of(&state)?;
    let http = client();
    let base = base(&server_url);
    let mut req = http
        .get(format!("{base}/api/v1/reports/libro-compras"))
        .bearer_auth(token.expose_secret());
    if let Some(p) = period.as_ref().filter(|s| !s.is_empty()) {
        req = req.query(&[("period", p)]);
    }
    let resp = req.send().await.map_err(conn_error)?;
    if !resp.status().is_success() {
        return Err(error_message(resp).await);
    }
    resp.json()
        .await
        .map_err(|e| format!("Respuesta de libro de compras inválida del servidor: {e}"))
}

/// GET `/api/v1/reports/iva?period=YYYY-MM` (Bearer, admin+) — débito − crédito.
#[tauri::command]
pub async fn iva_summary(
    state: State<'_, SessionState>,
    server_url: String,
    period: Option<String>,
) -> Result<IvaSummary, String> {
    let token = token_of(&state)?;
    let http = client();
    let base = base(&server_url);
    let mut req = http
        .get(format!("{base}/api/v1/reports/iva"))
        .bearer_auth(token.expose_secret());
    if let Some(p) = period.as_ref().filter(|s| !s.is_empty()) {
        req = req.query(&[("period", p)]);
    }
    let resp = req.send().await.map_err(conn_error)?;
    if !resp.status().is_success() {
        return Err(error_message(resp).await);
    }
    resp.json()
        .await
        .map_err(|e| format!("Respuesta de IVA inválida del servidor: {e}"))
}

/// PATCH `/api/v1/purchase-orders/{id}/factura` (Bearer, admin+) — capturar el
/// documento del proveedor. Sólo se envía lo que el operador tenga a mano;
/// devuelve el libro del período para refrescar sin segunda llamada.
#[tauri::command]
#[allow(clippy::too_many_arguments)]
pub async fn set_po_invoice(
    state: State<'_, SessionState>,
    server_url: String,
    id: String,
    folio: Option<String>,
    date: Option<String>,
    tipo: Option<i32>,
    neto: Option<String>,
    iva: Option<String>,
    total: Option<String>,
) -> Result<PurchaseBook, String> {
    let token = token_of(&state)?;
    let http = client();
    let base = base(&server_url);
    let mut body = serde_json::json!({});
    if let Some(v) = folio.filter(|s| !s.is_empty()) {
        body["folio"] = serde_json::Value::String(v);
    }
    if let Some(v) = date.filter(|s| !s.is_empty()) {
        body["date"] = serde_json::Value::String(v);
    }
    if let Some(v) = tipo {
        body["tipo"] = serde_json::Value::from(v);
    }
    for (k, v) in [("neto", neto), ("iva", iva), ("total", total)] {
        if let Some(s) = v.filter(|s| !s.is_empty()) {
            body[k] = serde_json::Value::String(s);
        }
    }
    let resp = http
        .patch(format!("{base}/api/v1/purchase-orders/{id}/factura"))
        .bearer_auth(token.expose_secret())
        .json(&body)
        .send()
        .await
        .map_err(conn_error)?;
    if !resp.status().is_success() {
        return Err(error_message(resp).await);
    }
    resp.json()
        .await
        .map_err(|e| format!("Respuesta de factura inválida del servidor: {e}"))
}
